web/base_viewer.js

```diff
@@ -581,21 +581,35 @@ class BaseViewer {
           if (pdfDocument.loadingParams.disableAutoFetch || pagesCount > 7500) {
             // XXX: Printing is semi-broken with auto fetch disabled.
             this._pagesCapability.resolve();
-            return;
+            return null;
           }
           let getPagesLeft = pagesCount - 1; // The first page was already loaded.
 
           if (getPagesLeft <= 0) {
             this._pagesCapability.resolve();
-            return;
+            return null;
           }
+
+          const xfaDrawPromises = [];
+
           for (let pageNum = 2; pageNum <= pagesCount; ++pageNum) {
             pdfDocument.getPage(pageNum).then(
               pdfPage => {
                 const pageView = this._pages[pageNum - 1];
                 if (!pageView.pdfPage) {
                   pageView.setPdfPage(pdfPage);
                 }
+                if (isPureXfa) {
+                  // Force the rendering of all pages (in case the pdf is
+                  // printed). It isn't really an issue since the document
+                  // has been fully layed out in worker in order to just
+                  // get the total page number. For now, in order to print
+                  // a filled form, we print the generated html as the user
+                  // can see it. It could lead to a memory issue in case
+                  // the document as a huge number of pages which is pretty
+                  // unlikely since normally XFA is used to render forms.
+                  xfaDrawPromises.push(pageView.draw());
```

Comment:
*I'm finding this, as a matter of principle, somewhat annoying/bad to have to do just to enable printing; especially given e.g. the cases mentioned below.*
Is it really not possible to have the print-functionality generate any non-rendered XFA pages, similarly to how it creates normal canvases during printing?

For one thing, this short-circuits the caching of pages in the viewer in ways that will lead to weird bugs, too much memory usage (in a longer document), an inability to clean-up any allocated resourced, and just overall *very* inconsistent state.
Note that an entire document was *not* intended to be rendered at the same time, with the exception of one page documents, and that this breaks a lot of assumptions elsewhere in the viewer code.